sound/soc/sof/ipc4-mtrace.c

```diff
@@ -424,9 +426,19 @@ static int ipc4_mtrace_enable(struct snd_sof_dev *sdev)
 	msg.primary |= SOF_IPC4_MOD_INSTANCE(SOF_IPC4_MOD_INIT_BASEFW_INSTANCE_ID);
 	msg.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_FW_PARAM_SYSTEM_TIME);
 
-	/* The system time is in usec, UTC, epoch is 1601-01-01 00:00:00 */
-	kt = ktime_add_us(ktime_get_real(), FW_EPOCH_DELTA * USEC_PER_SEC);
-	system_time = ktime_to_us(kt);
+	/* do calibration to get the latency beween timestamp generation and fw processing */
+	send_time = local_clock();
+	msg.data_size = sizeof(system_time);
+	msg.data_ptr = &system_time;
+	/* get system time from fw */
+	ret = iops->set_get_data(sdev, &msg, msg.data_size, false);
+	if (ret)
+		return ret;
+	reply_time = local_clock();
+	fw_receive_latency = (reply_time - send_time) / 2;
+
+	/* The system time is in nsec and aligned with syslog */
+	system_time =local_clock() + fw_receive_latency;
```

Comment:
I have zero understanding of this PR.

If you send the system time that's suitable for Linux, how does the firmware know about it?

Conversely, if hypothetically a driver from another OS tried to use the firmware, what would be the epoch selected?

Reply:
FW uses 64bits dsp clock counter to build time at resolution of micro seconds, and only requires timestamp sent by host in format of 64bits xxxxxxx micro second. It doesn't know which epoch driver uses and only do time process with 64bits clock counter.  **Only log parser care about which epoch is used since it will convert the 64bits number into the epoch domain**.  

No matter what epoch driver use,  driver should send time in format of 64bits micro seconds for FW, . To make debugging easy, **driver should select the epoch used by the host OS log system**.  The tool for decoding host OS syslog takes key role for epoch selection which should be adopted by fw logging.

In Linux, we use dmesg to check the kernel log, in format of "%x:%y", x stands for seconds, y for micro seconds and this is useful for debugging, So we should choose dmesg time stamp (local_clock), not windows epoch since it is hard to compare Linux syslog with windows epoch log in FW.  

In windows, windows driver should use the epoch adopted by windows OS(KeQuerySystemTime). Windows has a tool to convert 64bits timestamp into its epoch domain.  Both local_clock() and KeQuerySystemTime return 64bits nanosecond,  so convert them to micro second and send to FW.  Above all, FW can support any epoch, but log parser should convert time back to epoch domain.

There are many time counters in Linux system, I tried every one and finally debugged the dmesg to find local_clock is the one match with dmesg log.